README.md
# Ninja-Deploy
Ninja Deploy is a deployment tool for deploying software to endpoints managed by the NinjaOne RRM Solution and is intended for NinjaOne Customers only but released under GNU GENERAL PUBLIC LICENSE V3 License to comply with the PowerShell Deployment Toolkit License Requirments and Silent HQ. We ask that you do not compy our code if you are not a customer of the NinjaOne Platform.

# Disclaimer & License
We may change the licening when ever possible and legally allowed to do so and if applicable we Reserve All Rights to our code except for portions unenforable by GNU GENERAL PUBLIC LICENSE v3



changelog.txt
Change Logs

2022 - SEPT - 16 : Added Ninja Deploy Google Chrome Enterprise (most up to date / most recent)
2022 - SEPT - 16 : Added Ninja Deploy Google Drive (most up to date / most recent)


deploy/Bitwarden/Deploy-Bitwarden.ps1
<#
.SYNOPSIS
    This script performs the installation or uninstallation of Bitwarden.
    # LICENSE #
    PowerShell App Deployment Toolkit - Provides a set of functions to perform common application deployment tasks on Windows.
    Copyright (C) 2017 - Sean Lillis, Dan Cunningham, Muhammad Mashwani, Aman Motazedian.
    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
    You should have received a copy of the GNU Lesser General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
.DESCRIPTION
    The script is provided as a template to perform an install or uninstall of an application(s).
    The script either performs an "Install" deployment type or an "Uninstall" deployment type.
    The install deployment type is broken down into 3 main sections/phases: Pre-Install, Install, and Post-Install.
    The script dot-sources the AppDeployToolkitMain.ps1 script which contains the logic and functions required to install or uninstall an application.
.PARAMETER DeploymentType
    The type of deployment to perform. Default is: Install.
.PARAMETER DeployMode
    Specifies whether the installation should be run in Interactive, Silent, or NonInteractive mode. Default is: Interactive. Options: Interactive = Shows dialogs, Silent = No dialogs, NonInteractive = Very silent, i.e. no blocking apps. NonInteractive mode is automatically set if it is detected that the process is not user interactive.
.PARAMETER AllowRebootPassThru
    Allows the 3010 return code (requires restart) to be passed back to the parent process (e.g. SCCM) if detected from an installation. If 3010 is passed back to SCCM, a reboot prompt will be triggered.
.PARAMETER TerminalServerMode
    Changes to "user install mode" and back to "user execute mode" for installing/uninstalling applications for Remote Destkop Session Hosts/Citrix servers.
.PARAMETER DisableLogging
    Disables logging to file for the script. Default is: $false.
.EXAMPLE
    PowerShell.exe .\Deploy-Bitwarden.ps1 -DeploymentType "Install" -DeployMode "NonInteractive"
.EXAMPLE
    PowerShell.exe .\Deploy-Bitwarden.ps1 -DeploymentType "Install" -DeployMode "Silent"
.EXAMPLE
    PowerShell.exe .\Deploy-Bitwarden.ps1 -DeploymentType "Install" -DeployMode "Interactive"
.EXAMPLE
    PowerShell.exe .\Deploy-Bitwarden.ps1 -DeploymentType "Uninstall" -DeployMode "NonInteractive"
.EXAMPLE
    PowerShell.exe .\Deploy-Bitwarden.ps1 -DeploymentType "Uninstall" -DeployMode "Silent"
.EXAMPLE
    PowerShell.exe .\Deploy-Bitwarden.ps1 -DeploymentType "Uninstall" -DeployMode "Interactive"
.NOTES
    Toolkit Exit Code Ranges:
    60000 - 68999: Reserved for built-in exit codes in Deploy-Application.ps1, Deploy-Application.exe, and AppDeployToolkitMain.ps1
    69000 - 69999: Recommended for user customized exit codes in Deploy-Application.ps1
    70000 - 79999: Recommended for user customized exit codes in AppDeployToolkitExtensions.ps1
.LINK
    http://psappdeploytoolkit.com
#>
[CmdletBinding()]
Param (
    [Parameter(Mandatory=$false)]
    [ValidateSet('Install','Uninstall','Repair')]
    [string]$DeploymentType = 'Install',
    [Parameter(Mandatory=$false)]
    [ValidateSet('Interactive','Silent','NonInteractive')]
    [string]$DeployMode = 'Interactive',
    [Parameter(Mandatory=$false)]
    [switch]$AllowRebootPassThru = $false,
    [Parameter(Mandatory=$false)]
    [switch]$TerminalServerMode = $false,
    [Parameter(Mandatory=$false)]
    [switch]$DisableLogging = $false
)

Try {
    ## Set the script execution policy for this process
    Try { Set-ExecutionPolicy -ExecutionPolicy 'ByPass' -Scope 'Process' -Force -ErrorAction 'Stop' } Catch {}

    ##*===============================================
    ##* VARIABLE DECLARATION
    ##*===============================================
    ## Variables: Application
    [string]$appVendor = 'Bitwarden Inc.'
    [string]$appName = 'Bitwarden'
    [string]$appVersion = ''
    [string]$appArch = ''
    [string]$appLang = ''
    [string]$appRevision = ''
    [string]$appScriptVersion = '1.0.0'
    [string]$appScriptDate = 'XX/XX/20XX'
    [string]$appScriptAuthor = 'Jason Bergner'
    ##*===============================================
    ## Variables: Install Titles (Only set here to override defaults set by the toolkit)
    [string]$installName = ''
    [string]$installTitle = 'Bitwarden'

    ##* Do not modify section below
    #region DoNotModify

    ## Variables: Exit Code
    [int32]$mainExitCode = 0

    ## Variables: Script
    [string]$deployAppScriptFriendlyName = 'Deploy Application'
    [version]$deployAppScriptVersion = [version]'3.8.4'
    [string]$deployAppScriptDate = '26/01/2021'
    [hashtable]$deployAppScriptParameters = $psBoundParameters

    ## Variables: Environment
    If (Test-Path -LiteralPath 'variable:HostInvocation') { $InvocationInfo = $HostInvocation } Else { $InvocationInfo = $MyInvocation }
    [string]$scriptDirectory = Split-Path -Path $InvocationInfo.MyCommand.Definition -Parent

    ## Dot source the required App Deploy Toolkit Functions
    Try {
        [string]$moduleAppDeployToolkitMain = "$scriptDirectory\AppDeployToolkit\AppDeployToolkitMain.ps1"
        If (-not (Test-Path -LiteralPath $moduleAppDeployToolkitMain -PathType 'Leaf')) { Throw "Module does not exist at the specified location [$moduleAppDeployToolkitMain]." }
        If ($DisableLogging) { . $moduleAppDeployToolkitMain -DisableLogging } Else { . $moduleAppDeployToolkitMain }
    }
    Catch {
        If ($mainExitCode -eq 0){ [int32]$mainExitCode = 60008 }
        Write-Error -Message "Module [$moduleAppDeployToolkitMain] failed to load: `n$($_.Exception.Message)`n `n$($_.InvocationInfo.PositionMessage)" -ErrorAction 'Continue'
        ## Exit the script, returning the exit code to SCCM
        If (Test-Path -LiteralPath 'variable:HostInvocation') { $script:ExitCode = $mainExitCode; Exit } Else { Exit $mainExitCode }
    }

    #endregion
    ##* Do not modify section above
    ##*===============================================
    ##* END VARIABLE DECLARATION
    ##*===============================================

    If ($deploymentType -ine 'Uninstall' -and $deploymentType -ine 'Repair') {
        ##*===============================================
        ##* PRE-INSTALLATION
        ##*===============================================
        [string]$installPhase = 'Pre-Installation'

        ## Show Welcome Message, Close Bitwarden With a 60 Second Countdown Before Automatically Closing
        Show-InstallationWelcome -CloseApps 'Bitwarden' -CloseAppsCountdown 60

        ## Show Progress Message (with the default message)
        Show-InstallationProgress

        ## Uninstall Any Existing Version of Bitwarden
        $AppList = Get-InstalledApplication -Name 'Bitwarden'
        
        ForEach ($App in $AppList)
        {
        If($App.UninstallString)
        {
        $UninstPath = $App.UninstallString -replace '"', '' -replace '/allusers', ''
        
        If(Test-Path -Path $UninstPath)
        {
        Write-log -Message "Found $($App.DisplayName) ($($App.DisplayVersion)) and a valid uninstall string, now attempting to uninstall."

        Execute-Process -Path $UninstPath -Parameters '/allusers /S'
        Sleep -Seconds 5
        }
        }
        }
  
        ##*===============================================
        ##* INSTALLATION
        ##*===============================================
        [string]$installPhase = 'Installation'

        ## Install Bitwarden
        $ExePath = Get-ChildItem -Path "$dirFiles" -Include Bitwarden-Installer*.exe -File -Recurse -ErrorAction SilentlyContinue
        If($ExePath.Exists)
        {
        Write-Log -Message "Found $($ExePath.FullName), now attempting to install $installTitle."
        Show-InstallationProgress "Installing Bitwarden. This may take some time. Please wait..."
        Execute-Process -Path "$ExePath" -Parameters "/allusers /S" -WindowStyle Hidden

        ## Remove Bitwarden Desktop Shortcut
        if (Test-Path -Path "$envCommonDesktop\Bitwarden.lnk") {
        Write-Log -Message "Removing the Bitwarden Desktop Shortcut."
        Remove-Item -Path "$envCommonDesktop\Bitwarden.lnk" -Force -ErrorAction SilentlyContinue 
        }
        }
    
        ##*===============================================
        ##* POST-INSTALLATION
        ##*===============================================
        [string]$installPhase = 'Post-Installation'

    }
    ElseIf ($deploymentType -ieq 'Uninstall')
    {
        ##*===============================================
        ##* PRE-UNINSTALLATION
        ##*===============================================
        [string]$installPhase = 'Pre-Uninstallation'

        ## Show Welcome Message, Close Bitwarden With a 60 Second Countdown Before Automatically Closing
        Show-InstallationWelcome -CloseApps 'Bitwarden' -CloseAppsCountdown 60

        ## Show Progress Message (With a Message to Indicate the Application is Being Uninstalled)
        Show-InstallationProgress -StatusMessage "Uninstalling the $installTitle Application. Please Wait..."


        ##*===============================================
        ##* UNINSTALLATION
        ##*===============================================
        [string]$installPhase = 'Uninstallation'

        ## Uninstall Any Existing Version of Bitwarden
        $AppList = Get-InstalledApplication -Name 'Bitwarden'
        
        ForEach ($App in $AppList)
        {
        If($App.UninstallString)
        {
        $UninstPath = $App.UninstallString -replace '"', '' -replace '/allusers', ''
        
        If(Test-Path -Path $UninstPath)
        {
        Write-log -Message "Found $($App.DisplayName) ($($App.DisplayVersion)) and a valid uninstall string, now attempting to uninstall."

        Execute-Process -Path $UninstPath -Parameters '/allusers /S'
        Sleep -Seconds 5
        }
        }
        }

        ##*===============================================
        ##* POST-UNINSTALLATION
        ##*===============================================
        [string]$installPhase = 'Post-Uninstallation'


    }
    ElseIf ($deploymentType -ieq 'Repair')
    {
        ##*===============================================
        ##* PRE-REPAIR
        ##*===============================================
        [string]$installPhase = 'Pre-Repair'

        ## Show Progress Message (with the default message)
        Show-InstallationProgress


        ##*===============================================
        ##* REPAIR
        ##*===============================================
        [string]$installPhase = 'Repair'


        ##*===============================================
        ##* POST-REPAIR
        ##*===============================================
        [string]$installPhase = 'Post-Repair'


    }
    ##*===============================================
    ##* END SCRIPT BODY
    ##*===============================================

    ## Call the Exit-Script function to perform final cleanup operations
    Exit-Script -ExitCode $mainExitCode
}
Catch {
    [int32]$mainExitCode = 60001
    [string]$mainErrorMessage = "$(Resolve-Error)"
    Write-Log -Message $mainErrorMessage -Severity 3 -Source $deployAppScriptFriendlyName
    Show-DialogBox -Text $mainErrorMessage -Icon 'Stop'
    Exit-Script -ExitCode $mainExitCode
}


deploy/google/google_chromeenterprise.ps1
#Author: Christopher Sparrowgrove | Owner | CTO
#Company: IT Ninjas Technology, LLC
#Tool_Name: Ninja Deploy v. 3.0 - Ninja RMM Deploy Tool
#Script_Name: Ninja Deploy - Google Chrome
#Desc: NinjaRMM (NinjaOne) Silent Software Installer for Google Chrome Enterprise


#############################
##  APPLICATION VARIABLES  ##
#############################
$DeploymentSite = "https://YOUR_SITE_HERE.com/deploy";
$AppName = "Google_ChromeEnterprise";          $AppDir = "Google";          
$PackageHash = "";
$64bitPackage_URL = "https://dl.google.com/tag/s/appguid%3D%7B8A69D345-D564-463C-AFF1-A69D9E530F96%7D%26iid%3D%7B80A6A2D0-3C20-CA3B-DE10-5AF08F416431%7D%26lang%3Den%26browser%3D4%26usagestats%3D0%26appname%3DGoogle%2520Chrome%26needsadmin%3Dtrue%26ap%3Dx64-stable-statsdef_0%26brand%3DGCEA/dl/chrome/install/googlechromestandaloneenterprise64.msi";
$64bitPackageName = "googlechromestandaloneenterprise64.msi";
$32bitPackage_URL = "https://dl.google.com/tag/s/appguid%3D%7B8A69D345-D564-463C-AFF1-A69D9E530F96%7D%26iid%3D%7B80A6A2D0-3C20-CA3B-DE10-5AF08F416431%7D%26lang%3Den%26browser%3D4%26usagestats%3D0%26appname%3DGoogle%2520Chrome%26needsadmin%3Dtrue%26ap%3Dstable-arch_x86-statsdef_0%26brand%3DGCEA/dl/chrome/install/googlechromestandaloneenterprise.msi";
$32bitPackageName = "googlechromestandaloneenterprise.msi";
$PSADTK_URI = "https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/releases/download/3.8.4/PSAppDeployToolkit_v3.8.4.zip"; #RELEASE DATE: Jan 27, 2021
$PSADTK_FILE = "C:\ProgramData\NinjaRMMAgent\download\PSAppDeployToolkit\PSAppDeployToolkit_v3.8.4.zip";
$PSADTK_Dir = "C:\ProgramData\NinjaRMMAgent\download\PSAppDeployToolkit";

#############################
##   DEPLOYMENT VARIABLES  ##
#############################
$DeployTool = "Deploy-$AppName.ps1";       
$DeployTool_URL ="$DeploymentSite/$AppDir/$AppName/Deploy-$AppName.ps1";
$DeployDir = "C:\ProgramData\NinjaRMMAgent\download";
$LogsDir = "C:\ProgramData\NinjaRMMAgent\logs";     $TIMESTAMP = Get-Date -Format "MM/dd/yyyy HH:mm K";
$Logs = "C:\ProgramData\NinjaRMMAgent\logs\$AppName Deployment.log";


Write-Host "------------------------------------------------------------------------";


#PowerShell Deployment Tool Check
if(Get-Item -Path $PSADTK_FILE -ErrorAction Ignore) {Write-Output "[$TIMESTAMP] PowerShell AppDeployment Toolkit Check: Ok" >> $Logs; Write-Host "[$TIMESTAMP] PowerShell AppDeployment Toolkit (PSADTK) Check: Ok"}
else {
     New-Item -Path "C:\ProgramData\NinjaRMMAgent\download\PSAppDeployToolkit" -ItemType "directory"; Write-Output "[$TIMESTAMP] PowerShell AppDeployment Toolkit (PSADTK) Check: Directory Created" >> $Logs; Write-Host "[$TIMESTAMP] PowerShell AppDeployment Toolkit (PSADTK) Check: Directory Created";

    (New-Object System.Net.WebClient).DownloadFile($PSADTK_URI,$PSADTK_FILE); Write-Output "[$TIMESTAMP] PowerShell AppDeployment Toolkit  (PSADTK) ZipFile Check: Downloaded File" >> $Logs; Write-Host "[$TIMESTAMP] PowerShell AppDeployment Toolkit  (PSADTK) ZipFile Check: Downloaded File";
    
    #Unblock PSADTK
    Unblock-File -Path $PSADTK_FILE; Write-Output "PowerShell AppDeployment Toolkit Unblock: Unblocked" >> $Logs; Write-Host "[$TIMESTAMP] PowerShell AppDeployment Toolkit Unblock: Unblocked";

    #Extract PSADTK
    Expand-Archive -Path $PSADTK_FILE -Destination $PSADTK_Dir -Force; Write-Output "[$TIMESTAMP] PowerShell AppDeployment Toolkit  (PSADTK) Extraction Check: Extracted Zip File" >> $Logs; Write-Host "[$TIMESTAMP] PowerShell AppDeployment Toolkit  (PSADTK) Extraction Check: Extracted Zip File";
}

#Application Directory Check
if (Get-Item -Path $DeployDir"\"$AppDir"\"$AppName -ErrorAction Ignore) {Write-Host $OkMsg ; Write-Output "[$TIMESTAMP] $AppName Directory Check: OK" >> $Logs; Write-Host "[$TIMESTAMP] $AppName Directory Check: OK";}
else {New-Item -Path $DeployDir"\"$AppDir"\"$AppName -ItemType "directory"; Write-Host $CreatedMsg; Write-Output "[$TIMESTAMP] $$AppName Directory Check: Directory Created" >> $Logs;}

#AppDeploymentToolkit
if (Get-Item -Path $DeployDir"\"$AppDir"\$AppName\AppDeployToolkit" -ErrorAction Ignore) {Write-Output "[$TIMESTAMP] PSADTK Directory Check: OK" >> $Logs; Write-Host "[$TIMESTAMP] PSADTK Directory Check: OK";}
else {Copy-Item -Path $PSADTK_Dir"\Toolkit\AppDeployToolkit" -Destination $DeployDir"\"$AppDir"\"$AppName"\AppDeployToolkit" -Recurse;  Write-Output "[$TIMESTAMP] PSADTK Directory Check: Directory Created" >> $Logs; Write-Host  "[$TIMESTAMP] PSADTK Directory Check: Directory Created";}

# Files Directory
if (Get-Item -Path $DeployDir"\"$AppDir"\$AppName\Files" -ErrorAction Ignore) {Write-Output "[$TIMESTAMP] PSADTK Files Directory Check: Ok">> $Logs; Write-Host  "[$TIMESTAMP] PSADTK Files Directory Check: Ok";}
else {Copy-Item -Path $PSADTK_Dir"\Toolkit\Files" -Destination $DeployDir"\"$AppDir"\$AppName\Files" -Recurse;  Write-Output "[$TIMESTAMP] PSADTK Files Directory Check: Directory Created">> $Logs;} 

#Download Correct Software version
if ((gwmi win32_operatingsystem | select osarchitecture).osarchitecture -eq "64-bit")
    { 
      if (Get-Item -Path $DeployDir"\"$AppDir"\$AppName\Files\"$64bitPackageName -ErrorAction Ignore) {Write-Output "[$TIMESTAMP] $AppName 64bit File Check: Ok " >> $Logs;} 
      else {Invoke-WebRequest -Uri $64bitPackage_URL -Outfile $DeployDir"\"$AppDir"\$AppName\Files\"$64bitPackageName;  Write-Output "[$TIMESTAMP] $AppName 64bit File Check: Downloaded File " >> $Logs;} 
    }
else #Check if System is 32bit
    { 
        if(Get-Item -Path $DeployDir"\"$AppDir"\$AppName\Files\"$32bitPackageName -ErrorAction Ignore) {Write-Output "[$TIMESTAMP] $AppName 32bit File Check: Ok ">> $Logs;} 
        else {Invoke-WebRequest -Uri $32bitPackage_URL -Outfile $DeployDir"\"$AppDir"\$AppName\Files\"$32bitPackageName; Write-Output "[$TIMESTAMP] $AppName 32bit File Check: Downloaded File " >> $Logs;} 
    }

#Check Deploy Script
if(Get-Item -Path $DeployDir"\"$AppDir"\$AppName\"$DeployTool -ErrorAction Ignore) {Write-Output "[$TIMESTAMP] $AppName Deployment Script Check: Ok " >>  $Logs;} 
else {Invoke-WebRequest -Uri $DeployTool_URL -Outfile $DeployDir"\"$AppDir"\$AppName\"$DeployTool; Write-Output "[$TIMESTAMP] $AppName Deployment Script Check: Downloaded File " >>  $Logs;}

Powershell.exe -ExecutionPolicy Bypass "C:\ProgramData\NinjaRMMAgent\download\$AppDir\$AppName\$DeployTool" -DeploymentType "Install" -DeployMode "Silent"
Write-Output "[$TIMESTAMP] $AppName Deployment Started" >>  $Logs;


deploy/Google/Google_Drive/Deploy-Google_Drive.ps1
<#
.SYNOPSIS
    This script performs the installation or uninstallation of Google Drive.
    # LICENSE #
    PowerShell App Deployment Toolkit - Provides a set of functions to perform common application deployment tasks on Windows.
    Copyright (C) 2017 - Sean Lillis, Dan Cunningham, Muhammad Mashwani, Aman Motazedian.
    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
    You should have received a copy of the GNU Lesser General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
.DESCRIPTION
    The script is provided as a template to perform an install or uninstall of an application(s).
    The script either performs an "Install" deployment type or an "Uninstall" deployment type.
    The install deployment type is broken down into 3 main sections/phases: Pre-Install, Install, and Post-Install.
    The script dot-sources the AppDeployToolkitMain.ps1 script which contains the logic and functions required to install or uninstall an application.
.PARAMETER DeploymentType
    The type of deployment to perform. Default is: Install.
.PARAMETER DeployMode
    Specifies whether the installation should be run in Interactive, Silent, or NonInteractive mode. Default is: Interactive. Options: Interactive = Shows dialogs, Silent = No dialogs, NonInteractive = Very silent, i.e. no blocking apps. NonInteractive mode is automatically set if it is detected that the process is not user interactive.
.PARAMETER AllowRebootPassThru
    Allows the 3010 return code (requires restart) to be passed back to the parent process (e.g. SCCM) if detected from an installation. If 3010 is passed back to SCCM, a reboot prompt will be triggered.
.PARAMETER TerminalServerMode
    Changes to "user install mode" and back to "user execute mode" for installing/uninstalling applications for Remote Destkop Session Hosts/Citrix servers.
.PARAMETER DisableLogging
    Disables logging to file for the script. Default is: $false.
.EXAMPLE
    PowerShell.exe .\Deploy-GoogleDrive.ps1 -DeploymentType "Install" -DeployMode "Silent"
.EXAMPLE
    PowerShell.exe .\Deploy-GoogleDrive.ps1 -DeploymentType "Install" -DeployMode "Interactive"
.EXAMPLE
    PowerShell.exe .\Deploy-GoogleDrive.ps1 -DeploymentType "Uninstall" -DeployMode "Silent"
.EXAMPLE
    PowerShell.exe .\Deploy-GoogleDrive.ps1 -DeploymentType "Uninstall" -DeployMode "Interactive"
.NOTES
    Toolkit Exit Code Ranges:
    60000 - 68999: Reserved for built-in exit codes in Deploy-Application.ps1, Deploy-Application.exe, and AppDeployToolkitMain.ps1
    69000 - 69999: Recommended for user customized exit codes in Deploy-Application.ps1
    70000 - 79999: Recommended for user customized exit codes in AppDeployToolkitExtensions.ps1
.LINK
    http://psappdeploytoolkit.com
#>
[CmdletBinding()]
Param (
    [Parameter(Mandatory=$false)]
    [ValidateSet('Install','Uninstall','Repair')]
    [string]$DeploymentType = 'Install',
    [Parameter(Mandatory=$false)]
    [ValidateSet('Interactive','Silent','NonInteractive')]
    [string]$DeployMode = 'Interactive',
    [Parameter(Mandatory=$false)]
    [switch]$AllowRebootPassThru = $false,
    [Parameter(Mandatory=$false)]
    [switch]$TerminalServerMode = $false,
    [Parameter(Mandatory=$false)]
    [switch]$DisableLogging = $false
)

Try {
    ## Set the script execution policy for this process
    Try { Set-ExecutionPolicy -ExecutionPolicy 'ByPass' -Scope 'Process' -Force -ErrorAction 'Stop' } Catch {}

    ##*===============================================
    ##* VARIABLE DECLARATION
    ##*===============================================
    ## Variables: Application
    [string]$appVendor = 'Google'
    [string]$appName = 'Drive'
    [string]$appVersion = ''
    [string]$appArch = ''
    [string]$appLang = ''
    [string]$appRevision = ''
    [string]$appScriptVersion = '1.0.0'
    [string]$appScriptDate = 'XX/XX/20XX'
    [string]$appScriptAuthor = 'Jason Bergner'
    ##*===============================================
    ## Variables: Install Titles (Only set here to override defaults set by the toolkit)
    [string]$installName = ''
    [string]$installTitle = 'Google Drive'

    ##* Do not modify section below
    #region DoNotModify

    ## Variables: Exit Code
    [int32]$mainExitCode = 0

    ## Variables: Script
    [string]$deployAppScriptFriendlyName = 'Deploy Application'
    [version]$deployAppScriptVersion = [version]'3.8.4'
    [string]$deployAppScriptDate = '26/01/2021'
    [hashtable]$deployAppScriptParameters = $psBoundParameters

    ## Variables: Environment
    If (Test-Path -LiteralPath 'variable:HostInvocation') { $InvocationInfo = $HostInvocation } Else { $InvocationInfo = $MyInvocation }
    [string]$scriptDirectory = Split-Path -Path $InvocationInfo.MyCommand.Definition -Parent

    ## Dot source the required App Deploy Toolkit Functions
    Try {
        [string]$moduleAppDeployToolkitMain = "$scriptDirectory\AppDeployToolkit\AppDeployToolkitMain.ps1"
        If (-not (Test-Path -LiteralPath $moduleAppDeployToolkitMain -PathType 'Leaf')) { Throw "Module does not exist at the specified location [$moduleAppDeployToolkitMain]." }
        If ($DisableLogging) { . $moduleAppDeployToolkitMain -DisableLogging } Else { . $moduleAppDeployToolkitMain }
    }
    Catch {
        If ($mainExitCode -eq 0){ [int32]$mainExitCode = 60008 }
        Write-Error -Message "Module [$moduleAppDeployToolkitMain] failed to load: `n$($_.Exception.Message)`n `n$($_.InvocationInfo.PositionMessage)" -ErrorAction 'Continue'
        ## Exit the script, returning the exit code to SCCM
        If (Test-Path -LiteralPath 'variable:HostInvocation') { $script:ExitCode = $mainExitCode; Exit } Else { Exit $mainExitCode }
    }

    #endregion
    ##* Do not modify section above
    ##*===============================================
    ##* END VARIABLE DECLARATION
    ##*===============================================

    If ($deploymentType -ine 'Uninstall' -and $deploymentType -ine 'Repair') {
        ##*===============================================
        ##* PRE-INSTALLATION
        ##*===============================================
        [string]$installPhase = 'Pre-Installation'

        ## Show Welcome Message, Close Google Drive With a 60 Second Countdown Before Automatically Closing
        Show-InstallationWelcome -CloseApps 'GoogleDriveFS' -CloseAppsCountdown 60

        ## Show Progress Message (With a Message to Indicate the Application is Being Uninstalled)
        Show-InstallationProgress -StatusMessage "Removing Existing Versions of Google Drive. Please Wait..."

        ## Remove Existing Versions of Google Drive
        $AppList = Get-InstalledApplication -Name 'Google Drive'        
        ForEach ($App in $AppList)
        {
        If($App.UninstallString)
        {
        $UninstPath = $App.UninstallString -replace '"', ''       
        If(Test-Path -Path $UninstPath)
        {
        Write-log -Message "Found $($App.DisplayName) ($($App.DisplayVersion)) and a valid uninstall string, now attempting to uninstall."
        Execute-Process -Path $UninstPath -Parameters '--silent'
        Start-Sleep -Seconds 5
        }
        }
        }

        ## Check for Google Drive Pending Reboot
        If (Test-Path -Path "$envProgramFiles\Google\Drive File Stream\rebootpending") {
        Write-Log -Message "A system restart is required before the installation can proceed." -Severity 2
        Show-InstallationPrompt -Message "A system restart is required before the installation of $installTitle can proceed, please reboot at your earliest convenience." -ButtonRightText 'OK'
        Exit-Script -ExitCode 69004 #This code is to indicate a reboot is pending on this machine, and the installation cannot proceed.
        }
 
        ##*===============================================
        ##* INSTALLATION
        ##*===============================================
        [string]$installPhase = 'Installation'

        ## Install Google Drive for Desktop
        $ExePath = Get-ChildItem -Path "$dirFiles" -Include GoogleDrive.exe -File -Recurse -ErrorAction SilentlyContinue
        If($ExePath.Exists)
        {
        Write-Log -Message "Found $($ExePath.FullName), now attempting to install $installTitle."
        Show-InstallationProgress "Installing Google Drive for Desktop. This may take some time. Please wait..."
        Execute-Process -Path "$ExePath" -Parameters "--silent --gsuite_shortcuts=false" -WindowStyle Hidden
        Stop-Process -Name GoogleDriveFS -Force -ErrorAction SilentlyContinue
        }
       
        ##*===============================================
        ##* POST-INSTALLATION
        ##*===============================================
        [string]$installPhase = 'Post-Installation'

    }
    ElseIf ($deploymentType -ieq 'Uninstall')
    {
        ##*===============================================
        ##* PRE-UNINSTALLATION
        ##*===============================================
        [string]$installPhase = 'Pre-Uninstallation'

        ## Show Welcome Message, Close Google Drive With a 60 Second Countdown Before Automatically Closing
        Show-InstallationWelcome -CloseApps 'GoogleDriveFS' -CloseAppsCountdown 60

        ## Show Progress Message (With a Message to Indicate the Application is Being Uninstalled)
        Show-InstallationProgress -StatusMessage "Uninstalling the $installTitle Application. Please Wait..."

        ##*===============================================
        ##* UNINSTALLATION
        ##*===============================================
        [string]$installPhase = 'Uninstallation'

        ## Uninstall Existing Versions of Google Drive
        $AppList = Get-InstalledApplication -Name 'Google Drive'        
        ForEach ($App in $AppList)
        {
        If($App.UninstallString)
        {
        $UninstPath = $App.UninstallString -replace '"', ''       
        If(Test-Path -Path $UninstPath)
        {
        Write-log -Message "Found $($App.DisplayName) ($($App.DisplayVersion)) and a valid uninstall string, now attempting to uninstall."
        Execute-Process -Path $UninstPath -Parameters '--silent'
        Start-Sleep -Seconds 5
        }
        }
        }

        ##*===============================================
        ##* POST-UNINSTALLATION
        ##*===============================================
        [string]$installPhase = 'Post-Uninstallation'


    }
    ElseIf ($deploymentType -ieq 'Repair')
    {
        ##*===============================================
        ##* PRE-REPAIR
        ##*===============================================
        [string]$installPhase = 'Pre-Repair'

        ##*===============================================
        ##* REPAIR
        ##*===============================================
        [string]$installPhase = 'Repair'


        ##*===============================================
        ##* POST-REPAIR
        ##*===============================================
        [string]$installPhase = 'Post-Repair'


    }
    ##*===============================================
    ##* END SCRIPT BODY
    ##*===============================================

    ## Call the Exit-Script function to perform final cleanup operations
    Exit-Script -ExitCode $mainExitCode
}
Catch {
    [int32]$mainExitCode = 60001
    [string]$mainErrorMessage = "$(Resolve-Error)"
    Write-Log -Message $mainErrorMessage -Severity 3 -Source $deployAppScriptFriendlyName
    Show-DialogBox -Text $mainErrorMessage -Icon 'Stop'
    Exit-Script -ExitCode $mainExitCode
}
